# Blood faecal genomic comparison

This repository includes the code used to run the analysis in a recent study currently posted as a preprint on bioRxiv:

[Goodman, R.N., Moyo, S.J., Memelis, I., Khanijau, A., Manyahi, J., Kibwana, U.O., Aboud, S., Blomberg, B., Langeland, N. & Roberts, A.P. (2025) Genomic comparison of highly related pairs of *E. coli* and *K. pneumoniae* isolated from faeces and blood of the same neonatal patients hospitalized with fever in Dar es Salaam, Tanzania. bioRxiv. doi:10.1101/2025.04.28.650962.](https://doi.org/10.1101/2025.04.28.650962)

* [Figure 1](https://rngoodman.github.io/blood-faecal-genomic-comparison/vignettes/1_ANI_and_SNP_distance_matrices.html) - ANI and SNP distance matrices
* [Figure 2](https://rngoodman.github.io/blood-faecal-genomic-comparison/vignettes/2-core_genome_phylogenetic_tree.html) - core genome phylogenetic tree
* [Figure 3](https://rngoodman.github.io/blood-faecal-genomic-comparison/vignettes/3_variant_calling_heatmaps_with_breseq.html) - variant calling heatmaps with breseq
* [Figure 4](https://rngoodman.github.io/blood-faecal-genomic-comparison/vignettes/4_circular_whole_genome_comparisons.html) - circular whole genome comparisons
* [Figure 5](https://rngoodman.github.io/blood-faecal-genomic-comparison/vignettes/5_analysing_virulence_factor_determinants.html) - analysing virulence factor determinants
