# Blood Faecal Genomic comparisons

## Introduction 

This repository includes the code used to run the analysis in a recent mauuscript.

* [Figure 1](https://github.com/rngoodman/blob/main/vignettes/blood-faecal-genomic-comparison/5_analysing_virulence_factor_determinants.html) - ANI and SNP distance matrices
* [Figure 2](https://github.com/rngoodman/blood-faecal-genomic-comparison/vignettes/2-core_genome_phylogenetic_tree.html) - core genome phylogenetic tree
* [Figure 3](https://github.com/rngoodman/blood-faecal-genomic-comparison/vignettes/3_variant_calling_heatmaps_with_breseq.html) - variant calling heatmaps with breseq
* [Figure 4](https://github.com/rngoodman/blood-faecal-genomic-comparison/vignettes/4_circular_whole_genome_comparisons.html) - circular whole genome comparisons
* [Figure 5](https://github.com/rngoodman/blood-faecal-genomic-comparison/vignettes/5_analysing_virulence_factor_determinants.html) - analysing virulence factor determinants
