<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.55">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Richard Goodman">
<meta name="dcterms.date" content="2025-04-24">

<title>4: Circular Whole Genome Comparison</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="4_circular_whole_genome_comparisons_files/libs/clipboard/clipboard.min.js"></script>
<script src="4_circular_whole_genome_comparisons_files/libs/quarto-html/quarto.js"></script>
<script src="4_circular_whole_genome_comparisons_files/libs/quarto-html/popper.min.js"></script>
<script src="4_circular_whole_genome_comparisons_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="4_circular_whole_genome_comparisons_files/libs/quarto-html/anchor.min.js"></script>
<link href="4_circular_whole_genome_comparisons_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="4_circular_whole_genome_comparisons_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="4_circular_whole_genome_comparisons_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="4_circular_whole_genome_comparisons_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="4_circular_whole_genome_comparisons_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">4: Circular Whole Genome Comparison</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Richard Goodman </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 24, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>This tutorial will take two genome sequences and run a blastn comparison then visualise the comparison as a circlize output</p>
<p>This workflow involves using BLAST for sequence alignment, parsing the results, and visualizing the differences using circlize in R.</p>
<p>For a more in-depth tutorial on circular whole genome comparison see my <a href="https://github.com/rngoodman/circular-genome-comparisons">circular genome comparisons github repo</a>.</p>
</section>
<section id="part-1---genome-download-and-comparison-in-bash" class="level2">
<h2 class="anchored" data-anchor-id="part-1---genome-download-and-comparison-in-bash">Part 1 - Genome download and comparison in bash</h2>
<section id="download-reads-and-run-assemblies" class="level3">
<h3 class="anchored" data-anchor-id="download-reads-and-run-assemblies">1.1: Download reads and run assemblies</h3>
<p>See the <a href="https://rngoodman.github.io/blood-faecal-genomic-comparison/genome_assemblies.html">code on how to download reads and run assemblies</a></p>
</section>
<section id="download-reads-and-run-assemblies-1" class="level3">
<h3 class="anchored" data-anchor-id="download-reads-and-run-assemblies-1">1.2: Download reads and run assemblies</h3>
<p>You should now have the following files which we cna use for the tutorial</p>
<ul>
<li><code>FSBL1386_hybracter.fasta</code></li>
<li><code>FSFC1386_hybracter.fasta</code></li>
</ul>
</section>
<section id="rename-all-contig-headers-i.e.-everything-after-the" class="level3">
<h3 class="anchored" data-anchor-id="rename-all-contig-headers-i.e.-everything-after-the">1.3: Rename all contig headers i.e.&nbsp;everything after the “&gt;”</h3>
<p>Pattern breakdown:</p>
<ul>
<li>^(&gt;[^ ]+) → Captures the initial sequence identifier (e.g., &gt;NZ_CP102708.1).</li>
<li>.* (chromosome|plasmid [^,]+) → Captures the desired terms:</li>
<li>chromosome (single word).</li>
<li>plasmid [^,]+ (the full plasmid name before a comma or end of the line).</li>
<li><blockquote class="blockquote">
<p>\2 → Replaces the header with the extracted part.</p>
</blockquote></li>
</ul>
<p>Second sed step: * s/ /<em>/g → Replaces spaces with underscores. * s/^&gt;/&gt;9A-3-9</em>/ → Adds the prefix to all headers.</p>
<p>Note: this pattern is specific to these fasta files, use a different syntax for different headers</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Check fasta headings</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"^&gt;"</span> <span class="pp">*</span>.fasta</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># For single files</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-E</span> <span class="st">'s/^(&gt;[^ ]+) .* (chromosome|plasmid [^,]+).*/&gt;\2/'</span> FSBL1386.fasta <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-E</span> <span class="st">'s/ /_/g; s/^&gt;/&gt;FSBL1386_/'</span> <span class="op">&gt;</span> FSBL1386.fasta</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-E</span> <span class="st">'s/^(&gt;[^ ]+) .* (chromosome|plasmid [^,]+).*/&gt;\2/'</span> FSFC1386.fasta <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-E</span> <span class="st">'s/ /_/g; s/^&gt;/&gt;FSFC1386_/'</span> <span class="op">&gt;</span> FSFC1386.fasta</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># For multiple files</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> <span class="pp">*</span>.fasta<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="va">base</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="st">"</span><span class="va">$file</span><span class="st">"</span> .fasta<span class="va">)</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sed</span> <span class="at">-E</span> <span class="st">'s/^(&gt;[^ ]+) .* (chromosome|plasmid [^,]+).*/&gt;\2/'</span> <span class="st">"</span><span class="va">$file</span><span class="st">"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sed</span> <span class="at">-E</span> <span class="st">"s/ /_/g; s/^&gt;/&gt;</span><span class="va">${base}</span><span class="st">_/"</span> <span class="op">&gt;</span> <span class="st">"</span><span class="va">${base}</span><span class="st">.fasta"</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Check fasta headings again</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"^&gt;"</span> <span class="pp">*</span>.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="install-seqkit" class="level3">
<h3 class="anchored" data-anchor-id="install-seqkit">1.4: Install seqkit</h3>
<p>Make sure you have seqkit installed. You can download it from <a href="https://github.com/shenwei356/seqkit">Wei Shen’s Github page</a>.</p>
</section>
<section id="use-seqkit-to-get-contig-length" class="level3">
<h3 class="anchored" data-anchor-id="use-seqkit-to-get-contig-length">1.5: Use seqkit to get contig length</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> fx2tab <span class="at">-n</span> <span class="at">-l</span> FSBL1386_hybracter.fasta <span class="op">&gt;</span> FSBL1386_lengths.tsv</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> fx2tab <span class="at">-n</span> <span class="at">-l</span> FSFC1386_hybracter.fasta <span class="op">&gt;</span> FSFC1386_lengths.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="install-blast" class="level3">
<h3 class="anchored" data-anchor-id="install-blast">1.6: Install BLAST</h3>
<p>Make sure you have BLAST installed. You can download it from the <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI BLAST</a> website.</p>
</section>
<section id="use-blast-to-align-the-sequences" class="level3">
<h3 class="anchored" data-anchor-id="use-blast-to-align-the-sequences">1.7: Use BLAST to align the sequences</h3>
<p>Use <code>BLAST</code> to align the sequences in the FASTA files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define your reference and query genomes</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="va">REF</span><span class="op">=</span>FSFC1386_hybracter.fasta</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="va">Q</span><span class="op">=</span> <span class="ex">FSBL1386_hybracter.fasta</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Check your reference and query genomes</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="va">$REF</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="va">$Q</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create BLAST databases for both genomes</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="ex">makeblastdb</span> <span class="at">-in</span> <span class="va">${REF}</span> <span class="at">-dbtype</span> nucl <span class="at">-out</span> <span class="va">${REF}</span>_db</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="ex">makeblastdb</span> <span class="at">-in</span> <span class="va">${Q}</span> <span class="at">-dbtype</span> nucl <span class="at">-out</span> <span class="va">${Q}</span>_db</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform QAST to compare genome1 to genome2</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-query</span> <span class="va">${REF}</span> <span class="at">-db</span> <span class="va">${Q}</span>_db <span class="at">-out</span> <span class="va">${REF}</span>_vs_<span class="va">${Q}</span>.blastn <span class="at">-outfmt</span> 6</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform BLAST to compare genome2 to genome1</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="ex">blastn</span> <span class="at">-query</span> <span class="va">${Q}</span> <span class="at">-db</span> <span class="va">${REF}</span>_db <span class="at">-out</span> <span class="va">${Q}</span>_vs_<span class="va">${REF}</span>.blastn <span class="at">-outfmt</span> 6</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="part-2---visulisation-with-circulize-in-r" class="level2">
<h2 class="anchored" data-anchor-id="part-2---visulisation-with-circulize-in-r">Part 2 - Visulisation with circulize in R</h2>
<section id="parse-the-blast-results-to-get-the-alignments" class="level3">
<h3 class="anchored" data-anchor-id="parse-the-blast-results-to-get-the-alignments">2.1: Parse the BLAST results to get the alignments</h3>
<p>Parse BLAST Results</p>
<p>Parse the BLAST results to get the alignments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load necessary libraries</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(colorspace)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>parse_blast <span class="ot">&lt;-</span> <span class="cf">function</span>(blast_file) {</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  blast_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(blast_file, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(blast_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"qseqid"</span>, <span class="st">"sseqid"</span>, <span class="st">"pident"</span>, <span class="st">"length"</span>, <span class="st">"mismatch"</span>, </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"gapopen"</span>, <span class="st">"qstart"</span>, <span class="st">"qend"</span>, <span class="st">"sstart"</span>, <span class="st">"send"</span>, </span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"evalue"</span>, <span class="st">"bitscore"</span>)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(blast_data)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>blast_data <span class="ot">&lt;-</span> <span class="fu">parse_blast</span>(<span class="st">"BLAST_data/FSBL1386_vs_FSFC1386.blastn"</span>)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>blood <span class="ot">=</span> <span class="st">"FSBL1386"</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>faecal <span class="ot">=</span> <span class="st">"FSFC1386"</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(blast_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    qseqid                   sseqid  pident  length mismatch
1 FSBL1386_chromosome00001 FSFC1386_chromosome00001 100.000 3341929        0
2 FSBL1386_chromosome00001 FSFC1386_chromosome00001 100.000 1998734        0
3 FSBL1386_chromosome00001 FSFC1386_chromosome00001  98.864    5548       43
4 FSBL1386_chromosome00001 FSFC1386_chromosome00001  98.864    5548       43
5 FSBL1386_chromosome00001 FSFC1386_chromosome00001  99.661    5316       16
6 FSBL1386_chromosome00001 FSFC1386_chromosome00001  99.661    5316       16
  gapopen  qstart    qend  sstart    send evalue bitscore
1       0 1999234 5341162       1 3341929      0  6171000
2       0       1 1998734 3341930 5340663      0  3691000
3      18 5152335 5157877 2247379 2252911      0     9878
4      18 4246612 4252144 3153102 3158644      0     9878
5       2 5015767 5021081 2629611 2634925      0     9716
6       2 4628844 4634158 3016534 3021848      0     9716</code></pre>
</div>
</div>
</section>
<section id="get-genome-lengths-from-seqkit-table" class="level3">
<h3 class="anchored" data-anchor-id="get-genome-lengths-from-seqkit-table">2.2: Get genome lengths from seqkit table</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load necessary libraries</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Read contig lengths from the seqkit output</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>contig_lengths_ref <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"BLAST_data/FSBL1386_lengths.tsv"</span>, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>contig_lengths_q <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"BLAST_data/FSFC1386_lengths.tsv"</span>, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>contig_lengths <span class="ot">=</span> <span class="fu">rbind</span>(contig_lengths_ref, contig_lengths_q)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(contig_lengths) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chr"</span>, <span class="st">"length"</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the data</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(contig_lengths)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                                                                               chr
1                                                               FSBL1386_chromosome00001 len=5341162 circular=true
2  FSBL1386_plasmid00001 length=143420 plasmid_copy_number_short=0.9x plasmid_copy_number_long=0.49x circular=true
3                                                                             FSFC1386_chromosome00001 len=5340663
4 FSFC1386_plasmid00001 length=143423 plasmid_copy_number_short=0.56x plasmid_copy_number_long=3.34x circular=true
5   FSFC1386_plasmid00002 length=6415 plasmid_copy_number_short=11.89x plasmid_copy_number_long=0.0x circular=true
   length
1 5341162
2  143420
3 5340663
4  143423
5    6415</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Update genome length table with actual contig sizes</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>genome_lengths <span class="ot">&lt;-</span> contig_lengths <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">start =</span> <span class="dv">0</span>, <span class="at">end =</span> length) <span class="sc">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(chr, start, end)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Add dummy numeric column as before</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>value <span class="ot">&lt;-</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="rename-genome-lengths" class="level3">
<h3 class="anchored" data-anchor-id="rename-genome-lengths">2.3: Rename genome lengths</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename to make contigs simpler</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">"chromosome0000"</span>, <span class="st">"chr"</span>, genome_lengths<span class="sc">$</span>chr)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"plasmid0000"</span>, <span class="st">"p"</span>, genome_lengths<span class="sc">$</span>chr)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"FS"</span>, <span class="st">""</span>, genome_lengths<span class="sc">$</span>chr)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, genome_lengths<span class="sc">$</span>chr)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"1386"</span>, <span class="st">""</span>, genome_lengths<span class="sc">$</span>chr)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove anything after "len=" or "length="</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"(len(gth)?).*"</span>, <span class="st">""</span>, genome_lengths<span class="sc">$</span>chr)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>genome_lengths<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">d)</span><span class="sc">\\</span><span class="st">s"</span>, <span class="st">""</span>, genome_lengths<span class="sc">$</span>chr, <span class="at">perl =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Check final genome lengths table</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(genome_lengths<span class="sc">$</span>chr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BL chr1" "BL p1"   "FC chr1" "FC p1"   "FC p2"  </code></pre>
</div>
</div>
</section>
<section id="rename-blast-data" class="level3">
<h3 class="anchored" data-anchor-id="rename-blast-data">2.4: Rename BLAST data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename to make contigs simpler</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>qseqid <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">"chromosome0000"</span>, <span class="st">"chr"</span>, blast_data<span class="sc">$</span>qseqid)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>qseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"plasmid0000"</span>, <span class="st">"p"</span>, blast_data<span class="sc">$</span>qseqid)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>qseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"FS"</span>, <span class="st">""</span>, blast_data<span class="sc">$</span>qseqid)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>qseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, blast_data<span class="sc">$</span>qseqid)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>qseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"1386"</span>, <span class="st">""</span>, blast_data<span class="sc">$</span>qseqid)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># rename to make contigs simpler</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>sseqid <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">"chromosome0000"</span>, <span class="st">"chr"</span>, blast_data<span class="sc">$</span>sseqid)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>sseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"plasmid0000"</span>, <span class="st">"p"</span>, blast_data<span class="sc">$</span>sseqid)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>sseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"FS"</span>, <span class="st">""</span>, blast_data<span class="sc">$</span>sseqid)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>sseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, blast_data<span class="sc">$</span>sseqid)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>blast_data<span class="sc">$</span>sseqid <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"1386"</span>, <span class="st">""</span>, blast_data<span class="sc">$</span>sseqid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="filter-the-blast-results-based-on-percentage-identity-and-length" class="level3">
<h3 class="anchored" data-anchor-id="filter-the-blast-results-based-on-percentage-identity-and-length">2.5: Filter the BLAST results based on percentage identity and length</h3>
<p>For percentage identity: you can define a number e.g.&nbsp;99.5% is <code>f_pc_ident = 99.5</code></p>
<p>For minimum length: you can define number of bp e.g.&nbsp;1500bp is <code>f_length = 1500</code> or the % of the smallest contig e.g.&nbsp;60% of the smallest contig is <code>f_length = (min(60*(genome_lengths$end)))</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>f_pc_ident <span class="ot">=</span> <span class="fl">99.5</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>f_length <span class="ot">=</span> <span class="dv">1500</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>blast_data <span class="ot">=</span> blast_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pident <span class="sc">&gt;=</span> f_pc_ident <span class="sc">&amp;</span> length <span class="sc">&gt;</span> f_length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="adding-colour-to-plot" class="level3">
<h3 class="anchored" data-anchor-id="adding-colour-to-plot">2.6: Adding colour to plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define colors for specific sequences</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># here blood isolates are shades of red and faecal are shades of blue </span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate the genome lengths into two groups based on the "chr" name</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>blue_genomes <span class="ot">&lt;-</span> genome_lengths<span class="sc">$</span>chr[<span class="fu">grep</span>(<span class="st">"FC"</span>, genome_lengths<span class="sc">$</span>chr)]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>red_genomes <span class="ot">&lt;-</span> genome_lengths<span class="sc">$</span>chr[<span class="fu">grep</span>(<span class="st">"BL"</span>, genome_lengths<span class="sc">$</span>chr)]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign shades of blue for faecal genomes and shades of red for blood genomes</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>blue_colors <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">rainbow</span>(<span class="fu">length</span>(blue_genomes), <span class="at">start =</span> <span class="fl">0.6</span>, <span class="at">end =</span> <span class="fl">0.65</span>), blue_genomes)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>red_colors <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">rainbow</span>(<span class="fu">length</span>(red_genomes), <span class="at">start =</span> <span class="dv">0</span>, <span class="at">end =</span> <span class="fl">0.05</span>), red_genomes)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine the two color sets</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(blue_colors, red_colors)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the resulting colors</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  FC chr1     FC p1     FC p2   BL chr1     BL p1 
"#0066FF" "#0040FF" "#0019FF" "#FF0000" "#FF4D00" </code></pre>
</div>
</div>
</section>
<section id="visualize-with-circlize" class="level3">
<h3 class="anchored" data-anchor-id="visualize-with-circlize">2.7: Visualize with circlize</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Clear previous plots</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.clear</span>()</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.par</span>(<span class="at">gap.degree =</span> <span class="dv">5</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize the genome data</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.genomicInitialize</span>(genome_lengths, <span class="at">plotType =</span> <span class="st">"axis"</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a new track with different colors for each chromosome</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.genomicTrackPlotRegion</span>(genome_lengths, <span class="at">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  sector.index <span class="ot">&lt;-</span> <span class="fu">get.cell.meta.data</span>(<span class="st">"sector.index"</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  col <span class="ot">&lt;-</span> colors[sector.index]</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">circos.genomicRect</span>(region, value, <span class="at">col =</span> col, <span class="at">border =</span> <span class="cn">NA</span>)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>}, <span class="at">bg.border =</span> <span class="cn">NA</span>, <span class="at">track.height =</span> <span class="fl">0.1</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Add custom sector labels with rotation</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.track</span>(<span class="at">track.index =</span> <span class="dv">1</span>, <span class="at">panel.fun =</span> <span class="cf">function</span>(x, y) {</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  sector.index <span class="ot">&lt;-</span> <span class="fu">get.cell.meta.data</span>(<span class="st">"sector.index"</span>)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  xcenter <span class="ot">&lt;-</span> <span class="fu">get.cell.meta.data</span>(<span class="st">"xcenter"</span>)</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">circos.text</span>(</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> xcenter, </span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">get.cell.meta.data</span>(<span class="st">"ylim"</span>)[<span class="dv">1</span>], </span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> sector.index, </span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">facing =</span> <span class="st">"reverse.clockwise"</span>, </span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">niceFacing =</span> <span class="cn">TRUE</span>, </span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">adj =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.5</span>),</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex =</span> <span class="fl">0.6</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>}, <span class="at">bg.border =</span> <span class="cn">NA</span>)</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the connections with specific colors</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(blast_data)) {</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>  query_seq <span class="ot">&lt;-</span> blast_data<span class="sc">$</span>qseqid[i]</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>  color <span class="ot">&lt;-</span> colors[query_seq]</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">circos.genomicLink</span>(</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">region1 =</span> <span class="fu">data.frame</span>(<span class="at">chr =</span> blast_data<span class="sc">$</span>qseqid[i], <span class="at">start =</span> blast_data<span class="sc">$</span>qstart[i], <span class="at">end =</span> blast_data<span class="sc">$</span>qend[i]),</span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">region2 =</span> <span class="fu">data.frame</span>(<span class="at">chr =</span> blast_data<span class="sc">$</span>sseqid[i], <span class="at">start =</span> blast_data<span class="sc">$</span>sstart[i], <span class="at">end =</span> blast_data<span class="sc">$</span>send[i]),</span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">adjustcolor</span>(color, <span class="at">alpha.f =</span> <span class="fl">0.5</span>)</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a title to the plot with increased text size</span></span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="fu">paste0</span>(blood, <span class="st">" "</span>, <span class="st">"vs"</span>, <span class="st">" "</span>, faecal), <span class="at">cex.main =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="4_circular_whole_genome_comparisons_files/figure-html/circos_plot-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>